zebra-state/src/service.rs

```diff
@@ -29,14 +34,102 @@ struct StateService {
+}
+
+#[derive(Debug, Error)]
+#[error("block is not contextually valid")]
+struct CommitError(#[from] ValidateContextError);
+
+#[derive(displaydoc::Display, Debug, Error)]
+enum ValidateContextError {
+    /// block.height is lower than the current finalized height
+    OrphanedBlock,
 }
 
 impl StateService {
     pub fn new(config: Config, network: Network) -> Self {
         let sled = FinalizedState::new(&config, network);
-        let _mem = NonFinalizedState::default();
-        Self { sled, _mem }
+        let mem = NonFinalizedState::default();
+        let queued_blocks = QueuedBlocks::default();
+
+        Self {
+            sled,
+            mem,
+            queued_blocks,
+        }
+    }
+
+    #[instrument(skip(self, new))]
+    fn queue(&mut self, new: QueuedBlock) {
+        let parent_hash = new.block.header.previous_block_hash;
+
+        self.queued_blocks.queue(new);
+
+        if !self.contains(&parent_hash) {
+            return;
+        }
+
+        self.process_queued(parent_hash);
+
+        while self.mem.best_chain_len() > crate::constants::MAX_BLOCK_REORG_HEIGHT {
+            let finalized = self.mem.finalize();
+            self.sled
+                .commit_finalized_direct(finalized)
+                .expect("sled would never do us dirty like that");
+        }
+
+        self.queued_blocks
+            .prune_by_height(self.sled.finalized_tip_height());
+    }
+
+    fn validate_and_commit(&mut self, block: Arc<Block>) -> Result<(), CommitError> {
+        self.check_contextual_validity(&block)?;
+        let parent_hash = block.header.previous_block_hash;
+
+        if self.sled.finalized_tip_hash() == parent_hash {
+            self.mem.commit_new_chain(block);
+        } else {
+            self.mem.commit_block(block);
+        }
+
+        Ok(())
+    }
+
+    fn contains(&self, hash: &block::Hash) -> bool {
+        self.mem.any_chain_contains(hash) || &self.sled.finalized_tip_hash() == hash
+    }
+
+    #[instrument(skip(self))]
+    fn process_queued(&mut self, new_parent: block::Hash) {
+        let mut new_parents = vec![new_parent];
+
+        while let Some(parent) = new_parents.pop() {
+            let queued_children = self.queued_blocks.dequeue_children(parent);
+
+            for QueuedBlock { block, rsp_tx } in queued_children {
+                let hash = block.hash();
+                let result = self
+                    .validate_and_commit(block)
+                    .map(|()| hash)
+                    .map_err(Into::into);
+                let _ = rsp_tx.send(result);
+                new_parents.push(hash);
+            }
+        }
+    }
+
+    fn check_contextual_validity(&mut self, block: &Block) -> Result<(), ValidateContextError> {
+        use ValidateContextError::*;
+
+        if block.coinbase_height().unwrap() <= self.sled.finalized_tip_height() {
+            Err(OrphanedBlock)?;
+        }
+
+        // Draw the rest of the owl
+        Ok(())
```

Comment:
Let's use plain English descriptions, marked using `TODO`, so they're easy to search for.

If there's a ticket or a tracking issue, let's also add the ticket number.

```suggestion
        // TODO: contextual validation design and implementation
        Ok(())
```